Adding an example of QLearning for simple navigation

emailweixu opened the pull request.

emailweixu committed:
Adding an example of QLearning for simple navigation

bayesian commented on python/social_bot/util/replay_buffer.py:
```diff
@@ -0,0 +1,200 @@
+            if self._do_not_sample_flags[idx]:
+                continue
+            indices.append(idx)
+            is_weights.append(is_weight)
```
where does variable is_weight come from? should be weight?

emailweixu replied:
Fixed

bayesian commented on examples/train_simple_navigation.py:
```diff
@@ -0,0 +1,319 @@
+        obs = np.transpose(obs, [2, 0, 1])
+        action, q = agent.act(obs, reward)
+        control = [
+            0.02 * (action // options.action_discretize_levels),
```
use env.action_space here.. get rid of hard coded 0.02, also why it is different from 0.4 used in  f_action_feature = lambda _, action: (0.4 * (action // 6), 0.4 * (action % 6))

emailweixu replied:
This makes the force in the range of 0 - 0.2. Too big force make it very difficult to control. I changed the force to be -{0.1, -0.05, 0, 0.05, 0.1} is the new version

emailweixu replied:
the 0.4 in f_action_feature is to normalize action feature in the range of -[1, 1]

bayesian commented on examples/train_simple_navigation.py:
```diff
@@ -0,0 +1,319 @@
+import torch.optim as optim
+
+
+class Options(object):
```
Add comments about algorithms implemented here: double DQN with dueling and priority experience replay. 

also seems very close to baseline.deepq algorithm, how is it different?

emailweixu replied:
Added. Not much difference from baseline. I wrote it to make it a self-contained example and get more understanding of the detail

bayesian commented on examples/train_simple_navigation.py:
```diff
@@ -0,0 +1,319 @@
+            self._acting_net.eval()
+            with torch.no_grad():
+                q_values = self._acting_net.calc_q_values(input)
+            self._acting_net.train()
```
why it needs to train here?

emailweixu replied:
This make _acting_net in train mode. I'll move it learn()

bayesian commented on python/social_bot/util/segment_tree.py:
```diff
@@ -0,0 +1,87 @@
+import operator
+
+
+class SegmentTree(object):
```
since we already have dependencies on baseline.. can we just use implementation there.

emailweixu replied:
We will not depend on baseline in the future.

bayesian commented on python/social_bot/util/replay_buffer.py:
```diff
@@ -0,0 +1,200 @@
+            for feature, batch_feature in zip(features, batch_features):
+                batch_feature.append(feature)
+        for i, batch_feature in enumerate(batch_features):
+            batch_features[i] = np.vstack(batch_feature)
```
feels like there are easier ways to do 110 - 116 here with np functions.. tuple to list then vstack

bayesian commented on examples/train_simple_navigation.py:
```diff
@@ -0,0 +1,319 @@
+        num_input_channels = num_image_channels * (options.history_length + 1)
+        if options.f_action_feature is not None:
+            num_input_channels += len(
+                (options.f_action_feature)(0)) * options.history_length
```
this does not seem to match signature of lambda function:
f_action_feature = lambda _, action

emailweixu replied:
This is because f_action_feature is treated as a class method and '_' is for self.

emailweixu committed:
Ground texture and requires view the goal for positive reward

Addressed review comments

Also made other changes to train_simple_navigation:
1. LeakyReLUlu instead of ReLto prevent dead ReLU units
2. Priority for experience near the reward
3. Using average discounted reward as a base for Q value to allow faster convergence
4. Use nstep reward for updating Q values

bayesian commented on examples/train_simple_navigation.py:
```diff
@@ -0,0 +1,507 @@
+        return (1 - p) * self._options.prioritized_replay_gamma0
+
+    def _update_reward_dist(self):
+        i = len(self._replay_buffer) - 2
```
use option.history_length instead of hard coded 2.

emailweixu committed:
Some tweaks of simple_navigation environment.

Ending an episode if the agent move away from agent too far
Continue an episode if the agent reaches the goal, by moving the goal to a different location.
